lib/trino-hdfs/src/main/java/io/trino/filesystem/hdfs/HadoopPaths.java

```diff
@@ -14,36 +14,44 @@
 package io.trino.filesystem.hdfs;
 
 import com.google.common.base.VerifyException;
+import com.google.common.collect.ImmutableSet;
 import io.trino.filesystem.Location;
 import org.apache.hadoop.fs.Path;
 
 import java.net.URI;
 import java.net.URISyntaxException;
+import java.util.Set;
 
 public final class HadoopPaths
 {
+    private static final Set<String> S3_SCHEMES = ImmutableSet.of("s3", "s3a", "s3n");
     private HadoopPaths() {}
 
     public static Path hadoopPath(Location location)
     {
         // hack to preserve the original path for S3 if necessary
         String path = location.toString();
         Path hadoopPath = new Path(path);
-        if ("s3".equals(hadoopPath.toUri().getScheme()) && !path.equals(hadoopPath.toString())) {
-            return new Path(toPathEncodedUri(location));
+        if (hadoopPath.toUri().getScheme() != null) {
+            if (S3_SCHEMES.contains(hadoopPath.toUri().getScheme()) && !path.equals(hadoopPath.toString())) {
+                if (hadoopPath.toUri().getScheme().equals("s3")) {
+                    return new Path(toPathEncodedUri(location, true));
+                }
+                return new Path(toPathEncodedUri(location, false));
+            }
         }
         return hadoopPath;
     }
 
-    private static URI toPathEncodedUri(Location location)
+    private static URI toPathEncodedUri(Location location, boolean addFragment)
```

Comment:
we `addFragment` is needed?

Reply:
https://github.com/trinodb/trino/blob/984755f736f600b1fd4370b42617f07164b1bbfb/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/RegisterTableProcedure.java#L174

calls 

https://github.com/trinodb/trino/blob/984755f736f600b1fd4370b42617f07164b1bbfb/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java#L979

https://github.com/trinodb/trino/blob/984755f736f600b1fd4370b42617f07164b1bbfb/lib/trino-hdfs/src/main/java/io/trino/filesystem/hdfs/HdfsFileSystem.java#L232

As we know for location = `s3a://test-bucket-trino//dir1//dir2//sample_table2/metadata`

`new Path(toPathEncodedUri(location, false));` -> `s3a://test-bucket-trino/dir1/dir2/sample_table2/metadata`
`new Path(toPathEncodedUri(location, true));` -> `s3a://test-bucket-trino/dir1/dir2/sample_table2/metadata#/dir1//dir2//sample_table2/metadata`

If we go with existing code `new Path(toPathEncodedUri(location, true));` then the path `s3a://test-bucket-trino/dir1/dir2/sample_table2/metadata#/dir1//dir2//sample_table2/metadata` doesn't return any files while [listing  path](https://github.com/trinodb/trino/blob/984755f736f600b1fd4370b42617f07164b1bbfb/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java#L979) but files are there present in `s3a://test-bucket-trino/dir1/dir2/sample_table2/metadata`. 

However, while creating a table I didn't see any issues as`s3a://test-bucket-trino//dir1//dir2//sample_table2` explicit path given in create table is converted to `s3://test-bucket-trino/dir1//dir2//sample_table2/` (notice only single `/` before dir1)
when it complains about `Cannot create a table on a non-empty location: s3a://test-bucket-trino//dir1//dir2//sample_table2` it shows two // before dir1 but actually its single slash before dir1 at S3 level.

Anyways this could be another issue but I faced it while registering table with such path in HadoopFs. I'll remove this change as I am going to remove register table test and add a simplified test which creates a table in HadoopFs and can be queryable in NativeFs.

Reply:
Yeah, if we need special handling for `s3a/s3n` that seems odd at least... might affect native FS too.

Reply:
It wasn't specific to s3a/s3n but I added it for those cases as I wanted to have test for it and wasn't sure how it impacts all the usecases. Anyways its been removed for now.